### ISSUE TITLE
[ OOP ] rewrite the logic that manage the commits from github repositories

### ISSUE THREAD

**Issue by raduiman** [mozilla-releng/firefox-infra-changelog#438]
Rewrite the logic to get and manage the commits from git repositories using OOP classes and methods.

## Where to start:
Using the logic from the modular version ->/fic_modules/git.py, implement the functionality in OOP.

## Requirements
* use classes and methods
* Runs on Py3.X
* Document the code

**Comment by raduiman** [comment]
I've started to work on this. -> branch ' issue-438 '

**Comment by zsoltfay** [comment]
Updated this branch. Added 3 new methods.

### Extract commit data from an iterator
```python
    def extract_commit_data(self):
        for commit in self.repo_data.commits():
            self.commit_message = commit.message
            self.commit_date = commit.commit.author.get("date")
            self.commit_sha = commit.sha
            self.commit_author = commit.commit.author.get("name")
            self.commit_author_email = commit.commit.author.get("email")
            self.commit_url = commit.url
```

### Extract last_checked from a local json file
```python
    def last_checked(self, repo_name):
        return json.load(self.load(CHANGELOG_REPO_PATH, repo_name.lower() + ".json")).get("0").get("last_checked")
```

### Extract local_version from a local json file
```python
    def local_version(self, repo_name):
        return json.load(self.load(CHANGELOG_REPO_PATH, repo_name.lower() + ".json")).get("0").get("last_release").get("version")
```
========================================================

Please note that we are receiving a shortcommit object and that one does not return ```files_changed```

**Comment by raduiman** [comment]
Continued to implement the GIT part. All to work in progress is on branch [issue-438](https://github.com/mozilla-releng/firefox-infra-changelog/tree/issue-438)

Add the methods to get the commit information from Github repositories.                                 
                                                                                                        
 - def repo_selector() to check the repo type and redirect it to the proper method to get the commits   
 - def not_tag() to get the commits for the repositories with tag                                       
 - def build_puppet() to get the commits from build-puppet                                              
 - def tag() to get the commits from the repositories with tag                                          
 - def commit_keyword() to get the commits from the repositories with commit keywords                   
 - def local_version() to get the version from the local json files    (made by @zsoltfay )                                 
 - def last_check() to get the date time of the last check from the local json files   (made by @zsoltfay )                  
 - def construct_commit() to update the list of commits with the new commit                             
 - def store_data() to update the atributes from DataVault with information from the current commit     
 - def files() get the files affected by the commit                                                     
 - def get_url() to get the commit URL                                                                  
 - def get_author_email() to get the email of the author of the commit                                  
 - def get_author() to get the author of the commit                                                     
 - def get_date() to get the date of the commit                                                         
 - def get_message() to get the message of the commit                                                   
 - def get_sha() to get the SHA of the commit                                                           
 - def repo_files() to get the files which affect the infrastructure from the local json file           
 - def repo_team() to get the team/owner of the repository                                              
                                                                                                        
Add two main methods to iterate through the repositories and through commits:                           
 - def commit_iterator()                                                                                
 - def repo_iterator()                 

Add commit_files_changed, last_check, release_date and list_of_commits to FIC_DataVault

**Comment by raduiman** [comment]
Updates:

 Add the logic to save only the commits that affect the infrastructure using 'folders-to-check', 'keyword == deploy' and 'release version tag'

Branch: [Issue-438](https://github.com/mozilla-releng/firefox-infra-changelog/tree/issue-438) commit: https://github.com/mozilla-releng/firefox-infra-changelog/commit/4d2c1c3561d06eb37cb1a822f27341b972e920b0

**Comment by zsoltfay** [comment]
Recent updates:
- added exit code 12: Not a git or hg repository
- added 3 new entries to the data vault: repo_name, last_local_push_id, repo_type
- moved last_checked method to filehandler.py
- renamed the repo_type() method to extract_repo_type()
- created new static method: delete_element_0
- created a few new methods that handle the creation of element 0
- updated the update_json() method to write the new and old commits to json from newest to oldest

**Comment by bccrisan** [comment]
Recent updates:
- Renamed a few definitions:
  -  `delete_element_0()` to `remove_first_element()`
  - `add_element_0_hg()` to `add_first_element_hg()`
  - `add_element_0_git()` to `add_first_element_git()`

 - Changed the update_json() to save the data in the directory `data` by
default or if it's specified, it will use the specified directory.
-  Modified remove_first_element() method to not use a try-except, but instead made the method practically pop out the first key element from the dictionary and if it's 0, than remove it. Using `next(iter(some_dict))` removes the need of converting the dictionary
to a list (to get the first element from it) or to use iterators.

**Comment by raduiman** [comment]
Fixed in PR #448
